EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
-
-```
-(address, tree_key, sub_key)
-```
+The model makes a distinction between read and write events. As we will see later, gas costs are associated to these events.
 
-Where `tree_key` and `sub_key` are computed as `tree_key, sub_key = get_storage_slot_tree_keys(address, key)`
+#### Atomicity
 
-#### Access events for code
+The calls to `charge_gas()` are here to delimit the atomic boundaries of an operation. All the events emitted between two consecutive calls to `charge_gas` are charged together. If there isn't enough gas to pay for all events, none of the associated values are added to the witness.
 
-In the conditions below, “chunk chunk_id is accessed” is understood to mean an access event of the form
+#### Access events for account headers
 
-```
-(address, (chunk_id + 128) // 256, (chunk_id + 128) % 256)
+```python3
+# Emit all access events given an instruction. It is meant to be called before
+# the gas charging method for a given instruction is called, as these events
+# are subsequently used to process witness-related gas costs.
+def emit_access_events(opcode: int, stack: Stack, contract: Contract, new_contract: Optional[Contract]):
+  match opcode:
+    case OP_CALL | OP_DELEGATECALL | OP_STATICCALL | OP_CALLCODE | OP_SELFDESTRUCT:
+      address = stack.peek(1)
+
+      # XXX the previous version of the spec says OP_CALLCODE is concerned, but the geth
+      # code says it isn't.
+      if value_bearing and (op == OP_CALL or op == OP_SELFDESTRUCT or op == OP_CALLCODE):
+        add_write_access_event(contract.address, 0, BASIC_DATA_LEAF_KEY)
+        add_write_access_event(address, 0, BASIC_DATA_LEAF_KEY)
+
+        # If the operation triggers an account creation, charge the
+        # full cost of account creation
+        if not state.has_account(address):
+          add_write_access_event(address, 0, CODEHASH_LEAF_KEY)
+      else:
+        if is_precompile(address) or is_system_contract(address):
+          # XXX charge warm costs
+        else:
+          add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)
+       
+      add_read_access_event(contract.address, 0, BASIC_DATA_LEAF_KEY)
+
+    case OP_EXTCODECOPY | OP_EXTCODESIZE:
+      address = stack.peek(0)
+      add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)
+
+    case OP_BALANCE:
+      address = stack.peek(0)
+      add_read_access_event(address, 0, BASIC_DATA_LEAF_KEY)
+
+    case OP_CODECOPY:
+       if not contract.is_deployment:
+         add_read_access_event(contract.address, 0, BASIC_DATA_LEAF_KEY)
+
+    case OP_EXTCODEHASH:
+      address = stack.peek(0)
+      if not (is_precompile(address) or is_system_contract(address)):
+        add_read_access_event(address, 0, CODEHASH_LEAF_KEY)
+
+    case OP_SLOAD:
+      address = stack.peek(0)
+      key = stack.peek(1)
+      read_access_for_storage(address, key)
+
+    case OP_SSTORE:
+      address = stack.peek(0)
+      key = stack.peek(1)
+      write_access_for_storage(address, key)
```

Comment:
Same as above, contractAddress instead of stack.peek(0), and stack.peek(0) as key